systemtest/benchtest/main.go

```diff
@@ -232,22 +232,35 @@ func Run(allBenchmarks ...BenchmarkFunc) error {
 // warmup sends events to the remote APM Server using the specified number of
 // agents for the specified duration.
 func warmup(agents int, duration time.Duration, url, token string) error {
-	ctx, cancel := context.WithTimeout(context.Background(), duration)
-	defer cancel()
-
 	rl := loadgen.GetNewLimiter(loadgencfg.Config.MaxEPM)
+	h, err := newEventHandler(`*.ndjson`, url, token, rl)
+	if err != nil {
+		return fmt.Errorf("unable to create warm-up handler: %w", err)
+	}
 	var wg sync.WaitGroup
+	var once sync.Once
+	var ctx context.Context
+	var cancel context.CancelFunc
+	ctxReady := make(chan struct{})
 	for i := 0; i < agents; i++ {
-		h, err := loadgen.NewEventHandler(`*.ndjson`, url, token, rl)
-		if err != nil {
-			return fmt.Errorf("unable to create warm-up handler: %w", err)
-		}
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
+			// Create the context within the first goroutine that calls once.
+			// This ensures that the timeout starts counting from the moment
+			// the warm-up traffic starts.
+			// This delay starts becoming more noticeable once the number of
+			// agents increases.
```

Comment:
Did you make this change before you made the EventHandler safe for concurrent use? I'm surprised that creating the timeout context here vs. just before the loop makes a difference, seeing as there's practically nothing done at the beginning of the loop anymore.

Reply:
oh I forgot to remove this code. Thanks for catching this.